File: compiler/src/main/java/com/aaroncoplan/waterfall/compiler/statements/LambdaFunctionData.java
package com.aaroncoplan.waterfall.compiler.statements;

import com.aaroncoplan.waterfall.generated.WaterfallParser;
import com.aaroncoplan.waterfall.parser.Pair;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class LambdaFunctionData {

    public final List<Pair<String, String>> typedArguments;
    public final FunctionCallData body;  // null when the body is an empty `{}`

    public LambdaFunctionData(String filePath, WaterfallParser.LambdaFunctionContext ctx) {
        List<WaterfallParser.TypedArgumentContext> args = ctx.typedArgumentList() == null
                ? Collections.emptyList()
                : ctx.typedArgumentList().typedArgument();
        this.typedArguments = args.stream()
                .map(a -> new Pair<>(a.type().getText(), a.name.getText()))
                .collect(Collectors.toList());
        this.body = ctx.functionCall() == null ? null : new FunctionCallData(filePath, ctx.functionCall());
    }

    public String translate() {
        String argList = typedArguments.stream()
                .map(p -> p.firstVal + " " + p.secondVal)
                .collect(Collectors.joining(", "));
        String bodyText = body == null ? "{}" : body.translate();
        return String.format("(%s) ==> %s", argList, bodyText);
    }
}

File: compiler/src/main/kotlin/com/aaroncoplan/waterfall/compiler/statements/LambdaFunctionData.kt
package com.aaroncoplan.waterfall.compiler.statements

import com.aaroncoplan.waterfall.generated.WaterfallParser
import com.aaroncoplan.waterfall.parser.Pair

class LambdaFunctionData(filePath: String, ctx: WaterfallParser.LambdaFunctionContext) {

    @JvmField
    val typedArguments: List<Pair<String, String>> =
        (ctx.typedArgumentList()?.typedArgument() ?: emptyList()).map { a ->
            Pair(a.type().text, a.name.text)
        }

    /** Null when the source has an empty body `{}`. */
    @JvmField
    val body: FunctionCallData? =
        ctx.functionCall()?.let { FunctionCallData(filePath, it) }
}
